### PR TITLE
Bug/persistence schema [9021]

### PR DESCRIPTION
**Pull request by IkerLuengo**
Implementation and testing of the new persistence schema to store last sequence number as described in https://gitlab.intranet.eprosima.com/eProsima/fastrtps-internal-docs/blob/feature/dds-layer-persistence/design/persistence_schema_v2.0.md

### CODE REVIEW

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:37]: Check the return code of this function, it may fail.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:38]: The sentence `PRAGMA user_version;` should return a single result, I'd change `while` with `if`.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:43]: As `version_stmt` is local, we can omit this assignation.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp]: What do you think about moving this statement content to a global `static const std::string` at the top of the file?

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:94]: Are versions 0 and 1 the same?

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:121]: Why don't use the `SQLITE_OPEN_CREATE` flag directly it was as being done before?

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:172]: A global `static const std::string`?
Maybe even splitting the statements to be reused between other globals.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (LuisGP) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:330]:
```suggestion
        //First add the last seq number, it is needed for the foreign key on writers_histories
```

File: test/unittest/rtps/persistence/PersistenceTests.cpp
Comment (LuisGP) [inline test/unittest/rtps/persistence/PersistenceTests.cpp:199]: Reuse the statement from Fast-DDS to ease the maintenance of this test. Probably exposing somehow the create method directly.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [reply src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:94]: Versions prior to 2 are not defined. I was considering version '0' as the default return when no version pragma was on the database, and didn't like not having a version '1'.

Now I changed that, and version '1' is the default, and version '0' is used as an error signal when retrieving the version.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [reply src/cpp/rtps/persistence/SQLite3PersistenceService.cpp]: I like the idea of having all the upgrade methods self-contained. Keep in mind that future schema upgrades may be necessary (from 2 to 3, from 1 to 3...) and each needs a separate upgrade statement (some may need several statements with processing in between). I think it may look tidier on the long run.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [reply src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:121]: We want to avoid the upgrade on new databases. If the database does not exist and is created, it is created empty. there will be no schema version, so we will infer it is version '1', but the upgrade procedure will fail because there will be no tables on it.

I find this easier than trying a fail-free upgrade: if we try to open it without creating and it fails, but opening it with creation succeeds, it is a new one, will already be created with the appropriate schema, and we skip the upgrade (which is in the else clause)

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [reply src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:172]: I agree with the global in this case, as the database will always be created with the latest schema, so there will ever be a single statement to maintain. The splitting, on the other hand, will probably make the maintenance harder on the long run. The complete statements are easier to understand IMHO.

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [reply src/cpp/rtps/persistence/SQLite3PersistenceService.cpp]: I failed to find a maintainable way of doing this, even with macros (due to problems with escape characters + multiline strings and / or raw string literals.
We could use some method that returns the statements (for example, with the methods constructing the final string dynamically from stringstreams and static strings), but that would not fit on a header file and will not be easily shared with the test.

Any ideas?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2469/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2282/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2472/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2489/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2303/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2496/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3036/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/264/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2722/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3047/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3062/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/283/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2744/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3073/)

File: src/cpp/rtps/persistence/SQLite3PersistenceService.cpp
Comment (IkerLuengo) [inline src/cpp/rtps/persistence/SQLite3PersistenceService.cpp:220]: I'd make this an ERROR log, as we suspect it is the cause of the issues we are experiencing, and we already saw it may affect the user.

**Review by IkerLuengo — COMMENTED** [review]
Someone that was not involved in the development should review this PR

**Comment by jparisu** [comment]
It needs to be rebased to master before merge.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3076/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/293/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2754/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3094/)

**Review by imontesino — APPROVED** [review]
LGTM. Remaining errors seem to be related to another change unrelated to this PR.

**Pull request merged**
